### PR TITLE
Upload ZIP archive if one already exists in deployment folder

### PR DESCRIPTION
**Pull request by jhamet**
#### Which issue(s) does this change fix?
#4638 

#### Why is this change necessary?
A popular build system, [Pants](https://www.pantsbuild.org/docs/awslambda-python, creates zipped Lambda deployments automatically. We currently use this in conjunction with a custom `BuildMethod` via a `Makefile` for SAM for performant builds. The challenge right now is in order for `sam deploy` to function correctly, an unarchived directory needs to exist for SAM to then archive and upload. Overall, this results in the need to unzip in the `Makefile` implementation after the build just to get rezipped again which is the bottleneck in our build/deployments (especially as the number of Lambdas grow in a single SAM template).

#### How does it address the issue?
If there's a directory with a single zip file inside, avoid zipping.

#### What side effects does this change have?
None (that I am aware of).

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ✅ ] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [ ] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Comment by jhamet** [comment]
> Seems reasonable. Question on the Makefile, if the `makefile` is editable, could we change it such that it dumps artifacts in the ARTIFACTS_DIR and then we let `sam package` handle the zipping?

The problem in my case is the [Pants build system](https://www.pantsbuild.org/docs/awslambda-python) builds a zip for deployment (not configurable to be unzipped). Thus, users of this build system need to unzip into ARTIFACTS_DIR  if they want to use SAM for deployment (which then subsequently rezips upon package/deploy).

File: samcli/lib/package/utils.py [hunk 1]
```diff
@@ -176,8 +175,15 @@ def upload_local_artifacts(
...
-    # Or, pointing to a folder. Zip the folder and upload (zip_method is changed based on resource type)
     if is_local_folder(local_path):
+        files = os.listdir(local_path)
+        # Custom build methods may automatically produce a
+        # Zip.
+        if len(files) == 1:
+            archive_path = os.path.join(local_path, files[0])
+            if is_zip_file(archive_path):
+                return uploader.upload_with_dedup(archive_path)
+        # Or, pointing to a folder. Zip the folder and upload (zip_method is changed based on resource type)
```
Comment (mndeveci) [inline samcli/lib/package/utils.py:186]: Thanks for your contribution! Can we add some logging here to track this down later?

**Review by hawflau — CHANGES_REQUESTED** [review]
Thanks for the contribution. Could you please address the import format issues, and also add some integration test cases? Thanks

**Comment by jhamet** [comment]
> Thanks for the contribution. Could you please address the import format issues, and also add some integration test cases? Thanks

Yep sorry for the delay will get on it.

**Comment by lucashuy** [comment]
Hi, sorry about the delayed, looks like `shutils` was removed during the refactoring of imports, which is why the tests are failing. Did you need any help with adding an integration test to test this new case?

**Comment by jhamet** [comment]
> Hi, sorry about the delayed, looks like `shutils` was removed during the refactoring of imports, which is why the tests are failing. Did you need any help with adding an integration test to test this new case?

Apologies for the delay. Added back that mistake.

I am curious for some feedback on how to add an integration test for this new case. I took a gander at how the current set is constructed. Would it be https://github.com/aws/aws-sam-cli/blob/8ac861cad28bee327d2521a0ae5e591eae78c0ea/tests/integration/buildcmd/test_build_cmd.py#L1775 for example ? I presume I can "fake" a build such that it produces a Zip in build directory to test the functionality (otherwise I would need to hoist the Pants build system to the container running the test).

**Comment by jfuss** [comment]
@jhamet Are you still interesting in contributing this? If not, that is ok but would like to close it then. Happy to review an update if you are still interested in wrapping this up.

**Comment by hnnasit** [comment]
Hi @jhamet, since this PR has been inactive for a while I am going to close it. If you are still interested in contributing and wrapping the remaining changes, feel free to open a new PR with the changes requested and we will be happy to review the updates.

**Pull request closed without merging**